src/Grpc.AspNetCore.Server/Internal/HttpContextServerCallContext.cs

```diff
@@ -322,6 +342,15 @@ protected override Task WriteResponseHeadersAsyncCore(Metadata responseHeaders)
         // Clock is for testing
         public void Initialize(ISystemClock? clock = null)
         {
+            var currentActivity = Activity.Current;
+            if (currentActivity != null)
+            {
+                Activity.Current.AddTag("grpc.method", MethodCore);
+                RaiseActivityChanged(currentActivity);
```

Comment:
BTW, how do you know this is the activity ASP.NET Core created to track requests and not something created by user/other instrumentation site? Is it guaranteed?

Reply:
I think the code that interacts with DiagnosticSource will go away with this PR - https://github.com/aspnet/AspNetCore/pull/11685

Reply:
ok, what about AddTag? is it guaranteed to be called on proper Activity.Current?

Reply:
I don't know. What even would be the proper activity?

In my head I am imaging this will add tags to the request activity that is created by Kestrel host. Do you think it is better to check for the Kestrel host activity by name (also looking through parent activities)? The tags would then be added specifically to it rather than whatever the current activity happens to be. In that case, what should happen if Kestrel host activity is not present?

Reply:
I mean imagine the following
1. Kestrel (Hosting) start activity A. it becomes Current
2. User starts activity in the middleware B. It becomes Current (it may be a child of A or user may break it and it won't be a child of A)
3. We are here in this code now, grpc.method is set on activity B.
4. User stops B
5. Kestrel stops A: AppInsights/OpenTelemetry records Activity A - it never understands that it was grpc Activity because tags were set on B. This is not what we want, we always want tags on A. 
Without these tags on A AppInsights will show it as usual HTTP span/telemetry and it will be confusing.

Is there any place where user code executes and can create another Activity between Hosting and HttpContextServerCallContext.Initialize()? If there is, we should better create another grpc activity. 

BTW, If user code executes in async method - they will not modify this Activity.Current - only in their execution context, so may be we are good.

Reply:
The activity API is not easy enough for mere mortals to use. It’s actually insanely hard for us to get it right in the framework. Right now we have a single activity per request, are you saying we need one for each framework as well? Mvc, Grpc etc?

Reply:
I agree it's extremely hard to use and having two per request is not good. If there is a way to add tag on initial activity created by Kestrel - do it.

If there is a way to get reference to the instance of that activity without checking Current - this is the best option. 

If  Current is the only way - check if it's one from Kestrel and do not set tag if it's not.

Reply:
I've updated gRPC to only set tags on the host activity.